Module2

lishsicp opened the pull request.

lishsicp committed:
created entity classes using lombok

lishsicp committed:
Added simple DAO for tags:
    - dependency management
    - configured connection pool HikariCP for PostgreSQL driver
    - custom exception class with error codes

lishsicp committed:
Created simple Tag controller

lishsicp committed:
Added transaction support and profiles for db

lishsicp committed:
Added duplication exception, updates to TagDAO and TagService

lishsicp committed:
Validation exception handling for tag

lishsicp committed:
Merge branch 'tag-controller' into dev

lishsicp committed:
Added certificate dao
    - filter for certificates
    - moved exceptions to service
    - custom annotation for enums

lishsicp committed:
Changes to DAOs, exceptions now are thrown from dao

lishsicp committed:
Created Gift Certificate Service

lishsicp committed:
Created Gift Certificate Controller, added DuplicatedKeyException handle

lishsicp committed:
Added update validator for gift certificate

lishsicp committed:
Added ability to update gift certificate only with params that were passed in request

lishsicp committed:
Added i18n support for error messages

lishsicp committed:
Delete custom annotation

lishsicp committed:
Deleted useless enums

lishsicp committed:
Added config for embedded h2 tests for repo

lishsicp committed:
Changed delete method in TagServiceImpl

lishsicp committed:
Refactor validate method, added empty certificate check

lishsicp committed:
Added tagDao tests

lishsicp committed:
Changed getByName_ReturnsTagWithName test to check if tag is present

lishsicp committed:
Added getTagsForCertificates method

lishsicp committed:
Added giftCertificateDao tests

lishsicp committed:
Changed filtered search to receive http params

lishsicp committed:
Added builder to SearchFilter, changed variable names

lishsicp committed:
Code refactor

lishsicp committed:
updated exception handling to return errors in right format

lishsicp committed:
added one more test

lishsicp committed:
added test sql data

lishsicp committed:
added services tests

lishsicp committed:
Added the ability to pass additional info to DaoException
Changed update method to throw exception if certificate with id
do to exist

lishsicp committed:
Added new handlers. Added some java doc.

lishsicp committed:
Updated messages

lishsicp committed:
Fixed an issue when updating certificate
  if no tags specified during update all existing tags were removed
  from certificate

lishsicp committed:
test fix

lishsicp committed:
Added JavaDoc.
 also some minor code changes

lishsicp committed:
Merge branch 'certificate-controller' into dev

lishsicp committed:
Added data script

lishsicp committed:
Changed dev props

lishsicp committed:
Renamed classes

lishsicp committed:
Changed update query

lishsicp committed:
Match entries to RESTful convention

shelly-svg commented on model/pom.xml:
```diff
@@ -11,9 +11,40 @@
+        <dependency>
+            <groupId>com.zaxxer</groupId>
+            <artifactId>HikariCP</artifactId>
+            <version>5.0.1</version>
```
Please, move all the versions to the properties section same as you did for spring

shelly-svg commented on model/src/main/java/com/epam/esm/config/DBConfig.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Bean
+    public HikariConfig hikariConfig(
+            @Value("${spring.datasource.url}") String url,
```
Please, investiate @ConfigurationProperties

lishsicp replied:
thanks, ill consider it in 3rd module with boot

shelly-svg commented on model/src/main/java/com/epam/esm/dao/CRDDao.java:
```diff
@@ -0,0 +1,12 @@
+import java.util.List;
+
+public interface CRDDao<T> {
+    List<T> getAll();
```
Please, add javadocs to all interface methods everywhere

shelly-svg commented on model/src/main/java/com/epam/esm/exception/DaoException.java:
```diff
@@ -0,0 +1,26 @@
+package com.epam.esm.exception;
+
+public class DaoException extends Exception {
```
Why don`t we use RuntimeException?

shelly-svg commented on model/src/main/java/com/epam/esm/dao/GenericDao.java:
```diff
@@ -0,0 +1,13 @@
+import com.epam.esm.entity.Entity;
+import org.springframework.jdbc.core.JdbcTemplate;
+
+public abstract class GenericDao<T extends Entity> {
```
Why don`t we use Lombok?

shelly-svg commented on model/src/main/java/com/epam/esm/dao/QueryBuilder.java:
```diff
@@ -0,0 +1,60 @@
+public class QueryBuilder {
+
+    public String buildFilteredSelectQuery(String query, SearchFilter searchFilter) {
+        StringBuilder selectQuery = new StringBuilder(query);
```
Since we use Java 11 we can use "var"

shelly-svg commented on model/src/main/java/com/epam/esm/dao/QueryBuilder.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.Map;
+
+@Component
+public class QueryBuilder {
```
It would be better to create service interface and then implement it

shelly-svg commented on model/src/main/java/com/epam/esm/dao/QueryBuilder.java:
```diff
@@ -0,0 +1,60 @@
+@Component
+public class QueryBuilder {
+
+    public String buildFilteredSelectQuery(String query, SearchFilter searchFilter) {
```
Please, refactor this method. It is difficult to read it

shelly-svg replied:
Same for all methods at this class

shelly-svg replied:
Methods should be as short and as simple as possible. No more than 15 lines. You can divide this one

shelly-svg commented on model/src/main/java/com/epam/esm/dao/TagDao.java:
```diff
@@ -0,0 +1,17 @@
+
+    void assignTagToCertificate(Long certificateId, Long tagId);
+
+    void detachTagsFromCertificate(Long certificateId);
```
Why don`t we use simple types instead of wrappers?

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/GiftCertificateQuery.java:
```diff
@@ -0,0 +1,14 @@
+package com.epam.esm.dao.queries;
+
+public class GiftCertificateQuery {
```
I`d suggest to change the name of this class

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/GiftCertificateQuery.java:
```diff
@@ -0,0 +1,14 @@
+package com.epam.esm.dao.queries;
+
+public class GiftCertificateQuery {
+    private GiftCertificateQuery() {}
```
Add throws new UnsupportedOperationException() to the body of the private constructor

shelly-svg replied:
If you will want to use lombok, just use @UtilityClass

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/GiftCertificateQuery.java:
```diff
@@ -0,0 +1,14 @@
+    public static final String DELETE_BY_ID = "DELETE FROM gift_certificate WHERE id = ?";
+    public static final String UPDATE = "UPDATE gift_certificate SET ";
+    public static final String INSERT = "INSERT INTO gift_certificate (name, description, price, duration, create_date, last_update_date) VALUES (?,?,?,?,?,?)";
+    public static final String GET_ALL_FILTERED = "SELECT DISTINCT gc.id, gc.name, gc.description, gc.price, gc.duration, gc.create_date, gc.last_update_date FROM gift_certificate gc\n" +
```
Why LEFT JOIN?

lishsicp replied:
Left join works best here because otherwise certificates without tags wouldn't be displayed

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/TagQuery.java:
```diff
@@ -0,0 +1,14 @@
+    public static final String DELETE = "DELETE FROM tag WHERE id = ?";
+    public static final String INSERT = "INSERT INTO tag(name) VALUES(?)";
+    public static final String GET_TAGS_FOR_CERTIFICATE = "SELECT id, name FROM tag\n" +
+            "         JOIN gift_certificate_tag ON tag.id = gift_certificate_tag.tag_id\n" +
```
Why INNER JOIN?

lishsicp replied:
> Why INNER JOIN?

We can use any join here as null values are not possible in these tables, but if null values were possible then we wanna use inner join because we need to display only existing tags what were assigned to a specific certificate

shelly-svg commented on model/src/main/java/com/epam/esm/entity/Entity.java:
```diff
@@ -0,0 +1,6 @@
+
+import java.io.Serializable;
+
+public interface Entity extends Serializable {
```
I`d suggest to create some Abstract class let`s say "AbstractEntity" and place there this created_at updated_at fields. And extend this class at the all entities. ALso u can use @MappedSuperclass there

shelly-svg commented on model/src/main/java/com/epam/esm/exception/DaoExceptionErrorCode.java:
```diff
@@ -0,0 +1,20 @@
+package com.epam.esm.exception;
+
+public class DaoExceptionErrorCode {
```
Lets choose better name

shelly-svg commented on model/src/main/resources/data.sql:
```diff
@@ -0,0 +1,31 @@
+INSERT INTO tag (name)
```
Take a look on Liquibase library. It is widely used and 99% that you will have to deal with it on work. I`d suggest to use it in your project

lishsicp committed:
Added version props for dependencies in poms

lishsicp committed:
Added @UtilityClass annotation to query classes

lishsicp committed:
Renamed query classes

lishsicp committed:
Renamed ErrorCodes class

lishsicp committed:
Changed wrappers to primitives in dao

lishsicp committed:
Changed DaoException to unchecked

lishsicp committed:
Dao refactor

lishsicp committed:
QueryBuilder refactor

lishsicp committed:
Changed wrapper classes to primitives in service

lishsicp committed:
Changed IncorrectUpdateValueException to unchecked

lishsicp committed:
Changed dao test method names

lishsicp committed:
Changed service test method names and added some minor changes

lishsicp committed:
Added db migration using liquibase

lishsicp committed:
Added JavaDoc to interfaces

lishsicp committed:
Updated liquibase configuration files

lishsicp committed:
Updated dao test

lishsicp committed:
Refactor GiftCertificateQueryBuilder for improved readability

shelly-svg commented on model/src/main/java/com/epam/esm/dao/CRDDao.java:
```diff
@@ -0,0 +1,40 @@
+     * @param id Id of the entity to find
+     * @return Found entity
+     */
+    T getById(long id);
```
Please, read about Optional. And use it here

shelly-svg commented on model/src/main/java/com/epam/esm/dao/CRDDao.java:
```diff
@@ -0,0 +1,40 @@
+    void remove(long id);
+
+    /**
+     * Saves given entity, used when creating a new entity
```
Creates new entity

@param entityToCreate entity to create

@return created entity

shelly-svg commented on model/src/main/java/com/epam/esm/dao/impl/GiftCertificateDaoImpl.java:
```diff
@@ -0,0 +1,116 @@
+        Map<String, String> updateParams = getUpdateParams(giftCertificate);
+        String updateQuery = queryBuilder.buildUpdateQuery(CertificateQueries.UPDATE, updateParams);
+        int updatedRows = jdbcTemplate.update(updateQuery);
+        if (updatedRows == 0) throw new DaoException(ErrorCodes.CERTIFICATE_NOT_FOUND, giftCertificate.getId());
```
Please, always use
if (){
}

{} brackets are essential for readability, you are not supposed to use if without brackets event if there just one operator

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/CertificateQueries.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+public class CertificateQueries {
```
I`d suggest name - GiftCertificateSqlQueries

shelly-svg commented on model/src/main/java/com/epam/esm/dao/queries/CertificateQueries.java:
```diff
@@ -0,0 +1,15 @@
+
+@UtilityClass
+public class CertificateQueries {
+    public static final String GET_ALL = "SELECT * FROM gift_certificate";
```
Do you really need "static" here?

shelly-svg commented on model/src/main/java/com/epam/esm/dao/querybuilder/GiftCertificateQueryBuilder.java:
```diff
@@ -0,0 +1,55 @@
+@Component
+public class GiftCertificateQueryBuilder implements QueryBuilder {
+
+    public String buildFilteredSelectQuery(String selectQuery, SearchFilter searchFilter) {
```
Please, always use @Override when you override some method. It generally improves the readability of your code

shelly-svg commented on model/src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,45 @@
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
```
Please, use constructors with 4+ variables very carefully, its a bad practice

shelly-svg commented on model/src/main/java/com/epam/esm/entity/GiftCertificate.java:
```diff
@@ -0,0 +1,45 @@
+@Builder
+public class GiftCertificate implements Entity {
+
+  private long id;
```
It is bad practice to use auto-increment. Could you rework it to a better variant of id generation?
For the next time please prepare information and tell me:
1. What variants do we have and what pros-cons have each of database id generation?

shelly-svg commented on model/src/main/java/com/epam/esm/exception/DaoException.java:
```diff
@@ -0,0 +1,26 @@
+package com.epam.esm.exception;
+
+public class DaoException extends RuntimeException {
```
Please, read about Serialization and serialVersionUID

shelly-svg commented on model/src/main/resources/db/changelog/v-1.0/01-changeset-create-certificate-table.yaml:
```diff
@@ -0,0 +1,48 @@
+              - column:
+                  name: id
+                  type: bigint
+                  autoIncrement: true
```
Mentioned previously

shelly-svg commented on model/src/main/resources/db/changelog/v-1.0/01-changeset-create-certificate-table.yaml:
```diff
@@ -0,0 +1,48 @@
+databaseChangeLog:
+  - changeSet:
+      id: create-certificate
```
please, use more descriptive names, such as: "create gift certificate table structure", etc.

shelly-svg commented on model/src/main/resources/db/changelog/v-1.0/01-drop-tables.sql:
```diff
@@ -0,0 +1,9 @@
+alter table gift_certificate_tag
```
Do you need this sql files? If yes, move them to changelog

shelly-svg commented on model/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java:
```diff
@@ -0,0 +1,174 @@
+
+    @Test
+    void create_ShouldReturnCreatedCertificate() {
+        GiftCertificate certificate = new GiftCertificate();
```
Can you move such structures to some ModelFactory utility class that will contain all factory methods for such entities creation

shelly-svg commented on model/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java:
```diff
@@ -0,0 +1,174 @@
+    List<GiftCertificate> giftCertificateList;
+
+    @BeforeEach
+    void setUp() {
```
Please, rework your tests. You should avoid such constructions

shelly-svg commented on model/src/test/java/com/epam/esm/dao/impl/GiftCertificateDaoImplTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    void getAll_ShouldReturnAllCertificates_whenSearchFilterPassed() {
```
Did you check BDD, TDD? Please, rework your test structures

shelly-svg commented on service/src/main/java/com/epam/esm/service/exception/IncorrectUpdateValueException.java:
```diff
@@ -0,0 +1,14 @@
+
+public class IncorrectUpdateValueException extends RuntimeException {
+
+    final int errorCode;
```
why it is package-private?

shelly-svg commented on service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    @Override
+    public GiftCertificate findById(long id) {
```
whenever you are searching something by ID it is better to use Optional

shelly-svg commented on service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+
+        List<Tag> allTags = tagDao.getAll();
+
+        for (Tag newTag : newTags) {
```
Why dont you use something like

for (var tag : newTags){
tagService.findByName(tag.getName())
.ifPresentOrElse(...)
}

would be much readable

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ErrorBody.java:
```diff
@@ -0,0 +1,22 @@
+
+@Getter
+@Setter
+public class ErrorBody implements Serializable {
```
ErrorDto, dont need to be serializable btw

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ErrorBody.java:
```diff
@@ -0,0 +1,22 @@
+@Setter
+public class ErrorBody implements Serializable {
+
+    private String errorMessage;
```
Also can make it final

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ErrorBody.java:
```diff
@@ -0,0 +1,22 @@
+        this.errorCode = errorCode;
+    }
+
+    public ErrorBody() {
```
Do you need it?

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ExceptionControllerAdvice.java:
```diff
@@ -0,0 +1,129 @@
+ * @version 1.0
+ */
+@ControllerAdvice
+@Order(Ordered.HIGHEST_PRECEDENCE)
```
Why do you need it?

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ExceptionControllerAdvice.java:
```diff
@@ -0,0 +1,129 @@
+@ControllerAdvice
+@Order(Ordered.HIGHEST_PRECEDENCE)
+public class ExceptionControllerAdvice {
+
```
extra space

shelly-svg commented on web/src/main/java/com/epam/esm/exception/ExceptionControllerAdvice.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    /**
+     * Handles {@link DuplicateKeyException}
```
Dont need this javadocs, its excessive

lishsicp committed:
Refactor code and tests after PR review

- Move exception handling to service layer
- Use Optional when searching for entities
- Update Liquibase files
- General code refactor
- Update tests to use BDDMockito
- Remove custom validator
- Create ModelFactory class for entity creation

lishsicp committed:
Refactor code as well as some minor changes

lishsicp committed:
Moved Rest API Basic module to named folder

lishsicp committed:
Delete readme
